File: java/headFirstDesignPatterns/chap1/duckExample/src/br/com/duck/MallardDuck.java
package br.com.duck;

import br.com.duck.fly.FlyNoWay;
import br.com.duck.fly.FlyWithWings;
import br.com.duck.quack.Quack;

public class MallardDuck extends Duck{
    public MallardDuck(){
        quackBehavior = new Quack();
        flyBehavior = new FlyWithWings();
    }

    @Override
    public void notFly() {
        flyBehavior = new FlyNoWay();
    }
    
    
}

File: c_cpp/headFirst_designPatterns/chap1/duckExample/src/MallardDuck.cpp
/*
 * MallardDuck.cpp
 *
 */

#include "MallardDuck.hpp"

namespace ducks{

MallardDuck::MallardDuck(){
    quackBehavior = new quack::Quack();
    flyBehavior = new fly::FlyWithWings();
}

MallardDuck::~MallardDuck(){
    delete quackBehavior;
    delete flyBehavior;
}

void MallardDuck::notFly(){
    delete flyBehavior;
    flyBehavior = new fly::FlyNoWay();
}

};
